Make coarser triangulation with deal.II facilities

kronbichler opened the pull request:
This is due to extended functionality inside deal.II, mainly https://github.com/dealii/dealii/pull/12356 and the preparation steps before that. Note that I duplicate the function to coarsen from deal.II because I want to do different things on different levels.

This runs also several times faster than the old serial code.

kronbichler commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -420,124 +327,67 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
   });
 
   // for a single level nothing has to be done
-  if(fine_triangulation_in.n_global_levels() > 1)
-  {
-    auto const fine_triangulation =
-      dynamic_cast<parallel::distributed::Triangulation<dim, spacedim> const *>(
-        &fine_triangulation_in);
-
-    Assert(fine_triangulation, ExcNotImplemented());
```
Right now, I unconditionally delete the old code; if we still want to support dealii-9.3 (the last release version), which might be a good idea, we should wrap this into `#ifdef` clauses and switch between the two variants. (Noting that the newer one is faster and takes less memory.)

nfehn replied:
I thought dealii-9.3 is the new release? So did these changes come into dealii shortly after this year's release, or do you mean support of dealii-9.2?

peterrum replied:
We branched off the release branch before so these features are not part of 9.3.

nfehn commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -420,124 +327,67 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
+      8 * temp_triangulation.n_global_active_cells() / n_processes_last;
+    RepartitioningPolicyTools::MinimalGranularityPolicy<dim> policy(
+      std::min<unsigned int>(200, grain_size_limit));
+    auto const partition = policy.partition(temp_triangulation);
```
I think `partition` is not an intuitive name here (I am also not a fan of using `auto` in such cases as it makes the code difficult to understand; I guess the problem also originates from naming in deal.II since it is not intuitively clear to me what `policy.partition()` is doing; `partition` can be understood as noun or verb here). Here, I think of `partition` as one part of the triangulation, but it is then unclear to me why `partition.linfty_norm()` further below has something to do with the number of processes? Further below, I would guess `partition` takes the role of an mpi comm?

nfehn commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -420,124 +327,67 @@ create_geometric_coarsening_sequence(Triangulation<dim, spacedim> const & fine_t
+      std::min<unsigned int>(200, grain_size_limit));
+    auto const partition = policy.partition(temp_triangulation);
+    n_processes_last     = static_cast<unsigned int>(partition.linfty_norm() + 1);
+    partition.update_ghost_values();
```
Does it make a difference if changing the order of this line with the one above? If so, this would be a side-effect that needs to be documented. If not, I would place the line with the assignment of `n_processes_last = ...` somewhere else, e.g. at the end of the for-loop body, to indicate more clearly that this is just an update required for the next iteration of the for-loop.

nfehn replied:
I think you already changed this, so please resolve if this is the case.

kronbichler commented:
I now added `#if` guards for the 10.0 version of deal.II and the 9.3 release (and checked both ways locally), and rewrote the code significantly: I now add a new partitioning class that can be passed on to deal.II's function.

nfehn commented:
You could rebase so that the checks are run for 9.3 and 10.0

kronbichler commented:
Alright, it seems the CI does fine on both the 10.0 and 9.3 release.

nfehn commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -309,6 +312,54 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
+  }
+
+private:
+  mutable std::vector<types::global_cell_index> n_cells_per_level;
```
Why do you need this variable? It is private, but only filled in the function `partition()` without being used by this class or another one if I see it correctly.

kronbichler replied:
It is not used in the final version; however, it might be that we use more clever strategies in the future, so I left it in when I derived the algorithm in the present form. If you want, we can remove it now and bring it back in case we do a more refined strategy.

nfehn replied:
It does not help right now in my opinion.

nfehn commented on include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp:
```diff
@@ -309,6 +312,54 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
+    // TODO: We hard-code a grain-size limit of 200 cells per processor
+    // (assuming linear finite elements and typical behavior of
+    // supercomputers), but we also do not want to coarsen by more than a
+    // factor of 8 to not create too many MPI messages.
```
I don't understand the comment "coarsen by more than a factor of 8" completely (does it refer to the mesh or the MPI processes?). I would read the code like reducing the number of MPI processes by a factor of 8 for the next coarser level (at least as a target value). Maybe also information is missing here whether the vectors are filled from fine-to-coarse or coarse-to-fine.

kronbichler replied:
Updated comment in 4e6b0ff.

peterrum reviewed (approved):
Looks good to me :+1:
